File: P1411.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstring>
using namespace std;

const int MAXN = 705;

vector<int> graph[MAXN];
long long dp[MAXN][MAXN]; // dp[u][j]: 以u为根，包含u的连通块大小为j时，其他连通块的最大贡献
int sz[MAXN];
int n;

void dfs(int u, int parent) {
    sz[u] = 1;
    // 初始化：不切任何边时，包含u的连通块大小为1，其他连通块贡献为1（乘积单位元）
    dp[u][1] = 1;
    
    for (int v : graph[u]) {
        if (v == parent) continue;
        
        dfs(v, u);
        
        // 临时数组，用于更新dp值
        long long temp[MAXN];
        for (int j = 0; j <= sz[u]; j++) {
            temp[j] = dp[u][j];
        }
        
        // 更新子树大小
        int old_size = sz[u];
        sz[u] += sz[v];
        
        // 初始化新的dp值
        memset(dp[u], 0, sizeof(long long) * (sz[u] + 1));
        
        // 背包DP：考虑是否切断u-v边
        for (int j = old_size; j >= 1; j--) {
            if (temp[j] == 0) continue;
            
            // 情况1：切断u-v边，v子树独立
            // v子树的贡献是dp[v][0]（整个v子树的最大分数）
            long long cut_value = temp[j] * dp[v][0];
            if (cut_value > dp[u][j]) {
                dp[u][j] = cut_value;
            }
            
            // 情况2：不切断u-v边，合并连通块
            for (int k = 1; k <= sz[v]; k++) {
                if (dp[v][k] == 0) continue;
                
                // 合并后包含u的连通块大小为j+k
                long long merge_value = temp[j] * dp[v][k];
                if (j + k <= sz[u] && merge_value > dp[u][j + k]) {
                    dp[u][j + k] = merge_value;
                }
            }
        }
    }
    
    // 计算整个子树的最大分数：考虑包含u的连通块的贡献
    dp[u][0] = 0;
    for (int j = 1; j <= sz[u]; j++) {
        // 总分数 = 其他连通块贡献 * 包含u的连通块大小
        long long total = dp[u][j] * j;
        if (total > dp[u][0]) {
            dp[u][0] = total;
        }
    }
}

int main() {
    cin >> n;
    
    for (int i = 1; i < n; i++) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    
    // 初始化dp数组
    memset(dp, 0, sizeof(dp));
    
    dfs(1, 0);
    cout << dp[1][0] << endl;
    
    return 0;
}

File: P1411.py
import sys
sys.setrecursionlimit(10000)

def main():
    n = int(sys.stdin.readline().strip())
    graph = [[] for _ in range(n + 1)]
    
    # 读取边并构建图
    for _ in range(n - 1):
        u, v = map(int, sys.stdin.readline().split())
        graph[u].append(v)
        graph[v].append(u)
    
    # dp[u][j] 表示以u为根的子树中，除了包含u的大小为j的连通块之外的其他连通块的最大贡献
    # dp[u][0] 存储整个子树的最大分数
    dp = [[0] * (n + 1) for _ in range(n + 1)]
    size = [0] * (n + 1)
    
    def dfs(u, parent):
        size[u] = 1
        # 初始化：不切任何边时，包含u的连通块大小为1，其他连通块贡献为1（乘积单位元）
        dp[u][1] = 1
        
        for v in graph[u]:
            if v == parent:
                continue
                
            dfs(v, u)
            
            # 临时数组，用于更新dp值
            temp = [0] * (size[u] + 1)
            for j in range(size[u] + 1):
                temp[j] = dp[u][j]
            
            # 更新子树大小
            old_size = size[u]
            size[u] += size[v]
            
            # 初始化新的dp值
            for j in range(size[u] + 1):
                dp[u][j] = 0
            
            # 背包DP：考虑是否切断u-v边
            # 从大到小遍历，避免重复计算
            for j in range(old_size, 0, -1):
                if temp[j] == 0:
                    continue
                
                # 情况1：切断u-v边，v子树独立
                # v子树的贡献是dp[v][0]（整个v子树的最大分数）
                cut_value = temp[j] * dp[v][0]
                if cut_value > dp[u][j]:
                    dp[u][j] = cut_value
                
                # 情况2：不切断u-v边，合并连通块
                for k in range(1, size[v] + 1):
                    if dp[v][k] == 0:
                        continue
                    
                    # 合并后包含u的连通块大小为j+k
                    merge_value = temp[j] * dp[v][k]
                    if j + k <= size[u] and merge_value > dp[u][j + k]:
                        dp[u][j + k] = merge_value
        
        # 计算整个子树的最大分数：考虑包含u的连通块的贡献
        dp[u][0] = 0
        for j in range(1, size[u] + 1):
            # 总分数 = 其他连通块贡献 * 包含u的连通块大小
            total = dp[u][j] * j
            if total > dp[u][0]:
                dp[u][0] = total
    
    dfs(1, 0)
    print(dp[1][0])

if __name__ == "__main__":
    main()